ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/RemoteDataAugmenter.java

```diff
@@ -0,0 +1,157 @@
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.execution;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.rest.SessionProperties;
+import io.confluent.ksql.rest.client.RestResponse;
+import io.confluent.ksql.rest.entity.KsqlEntity;
+import io.confluent.ksql.rest.entity.KsqlEntityList;
+import io.confluent.ksql.rest.server.ServerUtil;
+import io.confluent.ksql.rest.util.DiscoverRemoteHostsUtil;
+import io.confluent.ksql.services.SimpleKsqlClient;
+import io.confluent.ksql.util.KsqlRequestConfig;
+import io.confluent.ksql.util.Pair;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import org.apache.kafka.streams.state.HostInfo;
+
+
+public final class RemoteDataAugmenter {
+  private static final int TIMEOUT_SECONDS = 10;
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteDataAugmenter.class);
+
+  private final String statementText;
+  private final SessionProperties sessionProperties;
+  private final KsqlExecutionContext executionContext;
+  private final SimpleKsqlClient ksqlClient;
+
+
+  private RemoteDataAugmenter(
+      final String statementText,
+      final SessionProperties sessionProperties,
+      final KsqlExecutionContext executionContext,
+      final SimpleKsqlClient ksqlClient
+  ) {
+    this.statementText = Objects.requireNonNull(statementText);
+    this.sessionProperties = Objects.requireNonNull(sessionProperties);
+    this.executionContext = Objects.requireNonNull(executionContext);
+    this.ksqlClient = Objects.requireNonNull(ksqlClient);
+    ;
+  }
+
+  public static RemoteDataAugmenter create(
+      final String statementText,
+      final SessionProperties sessionProperties,
+      final KsqlExecutionContext executionContext,
+      final SimpleKsqlClient ksqlClient
+  ) {
+    return new RemoteDataAugmenter(
+        statementText, sessionProperties, executionContext, ksqlClient);
+  }
+
+
+  public <R> R augmentWithRemote(
```

Comment:
We do want to return, reason being it might return some intermediary value that you want to do more work on rather than mutate. It's harder to unit test mutations than pure functions

Reply:
sure that sounds reasonable, but the way we're using it ListQueries is like a void function which is confusing, so my preference would be to actually capture the result of `augmentWithRemote()` in ListQueries and use that result in the rest of the function where augment is called.

Reply:
Done that, see other comment!